polus-apply-flatfield-plugin/src/main.py

```diff
@@ -1,11 +1,125 @@
+    
+        with BioReader(brightfield,max_workers=2) as bf:
+            brightfield_image = bf[:,:,:,0,0].squeeze()
+            
+        with BioReader(darkfield,max_workers=2) as df:
+            darkfield_image = df[:,:,:,0,0].squeeze()
+        
+        threads = []
+
+        for file in files:
+            
+            if photobleach != None:
+                pb = photo_offset[file['file']]
+            else:
+                pb = None
+            
+            ProcessManager.submit_thread(unshade_image,file['file'],
+                                                       out_dir,
+                                                       brightfield_image,
+                                                       darkfield_image,
+                                                       pb,
+                                                       offset)
+        
+        ProcessManager.join_threads()
+        
+def main(imgDir: Path,
+         imgPattern: str,
+         ffDir: Path,
+         brightPattern: str,
+         outDir: Path,
+         darkPattern: typing.Optional[str] = None,
+         photoPattern: typing.Optional[str] = None
+         ) -> None:
+    
+    ''' Start a process for each set of brightfield/darkfield/photobleach patterns '''
+    # Create the FilePattern objects to handle file access
+    ff_files = FilePattern(ffDir,brightPattern)
+    fp = FilePattern(imgDir,imgPattern)
+    if darkPattern != None and darkPattern!='':
+        dark_files = FilePattern(ffDir,darkPattern)
+    if photoPattern != None and photoPattern!='':
+        photo_files = FilePattern(str(Path(ffDir).parents[0].joinpath('metadata').absolute()),photoPattern)
+        
+    group_by = [v for v in fp.variables if v not in ff_files.variables]
+    GROUPED = group_by + ['file']
+
+    ProcessManager.init_processes('main','unshade')
+        
+    for files in fp(group_by=group_by):
+        
+        flat_path = ff_files.get_matching(**{k.upper():v for k,v in files[0].items() if k not in GROUPED})[0]['file']
+        if flat_path is None:
+            logger.warning("Could not find a flatfield image, skipping...")
+            continue
+        
+        if darkPattern is not None and darkPattern != '':
+            dark_path = dark_files.get_matching(**{k.upper():v for k,v in files[0].items() if k not in GROUPED})[0]['file']
+            
+            if dark_path is None:
+                logger.warning("Could not find a darkfield image, skipping...")
+                continue
+        
+        if photoPattern is not None and photoPattern != '':
+            photo_path = photo_files.get_matching(**{k.upper():v for k,v in files[0].items() if k not in GROUPED})[0]['file']
+            
+            if photo_path is None:
+                logger.warning("Could not find a photobleach file, skipping...")
+                continue
+            
+        ProcessManager.submit_process(unshade_batch,files,outDir,flat_path,dark_path)
```

Comment:
missing input: `photo_path`, plugin won't do photobleach correction even if the user wants to